op-plasma/damgr.go

```diff
@@ -2,48 +2,420 @@ package plasma
 }
 
 // NewPlasmaDA creates a new PlasmaDA instance with the given log and CLIConfig.
-func NewPlasmaDA(log log.Logger, cfg CLIConfig) *DA {
+func NewPlasmaDA(log log.Logger, cli CLIConfig, cfg Config, l1f L1Fetcher, metrics Metricer) *DA {
+	return NewPlasmaDAWithStorage(log, cfg, cli.NewDAClient(), l1f, metrics)
+}
+
+// NewPlasmaDAWithStorage creates a new PlasmaDA instance with the given log and DAStorage interface.
+func NewPlasmaDAWithStorage(log log.Logger, cfg Config, storage DAStorage, l1f L1Fetcher, metrics Metricer) *DA {
 	return &DA{
 		log:     log,
-		storage: cfg.NewDAClient(),
+		cfg:     cfg,
+		storage: storage,
+		l1:      l1f,
+		metrics: metrics,
+		state:   NewState(log, metrics),
 	}
 }
 
-// NewPlasmaDAWithStorage creates a new PlasmaDA instance with the given log and DAStorage interface.
-func NewPlasmaDAWithStorage(log log.Logger, storage DAStorage) *DA {
+// NewPlasmaWithState creates a plasma storage from initial state used for testing in isolation.
+func NewPlasmaDAWithState(log log.Logger, cfg Config, storage DAStorage, l1f L1Fetcher, metrics Metricer, state *State) *DA {
 	return &DA{
 		log:     log,
+		cfg:     cfg,
 		storage: storage,
+		l1:      l1f,
+		metrics: metrics,
+		state:   state,
 	}
 }
 
+// OnFinalizedHeadSignal sets the callback function to be called when the finalized head is updated.
+// This will signal to the engine queue that will set the proper L2 block as finalized.
+func (d *DA) OnFinalizedHeadSignal(f HeadSignalFn) {
+	d.finalizedHeadSignalFunc = f
+}
+
+// Finalize takes the L1 finality signal, compares the plasma finalized block and forwards the finality
+// signal to the engine queue based on whichever is most behind.
+func (d *DA) Finalize(l1Finalized eth.L1BlockRef) {
+	ref := d.finalizedHead
+	d.log.Info("received l1 finalized signal, forwarding to engine queue", "l1", l1Finalized, "plasma", ref)
+	// if the l1 finalized head is behind it is the finalized head
+	if l1Finalized.Number < d.finalizedHead.Number {
+		ref = l1Finalized
+	}
+	// prune finalized state
+	d.state.Prune(ref.Number)
+
+	if d.finalizedHeadSignalFunc == nil {
+		d.log.Warn("finalized head signal function not set")
+		return
+	}
+
+	// signal the engine queue
+	d.finalizedHeadSignalFunc(ref)
+}
+
+// LookAhead increments the challenges origin and process the new block if it exists.
+// It is used when the derivation pipeline stalls due to missing data and we need to continue
+// syncing challenge events until the challenge is resolved or expires.
+func (d *DA) LookAhead(ctx context.Context) error {
+	blkRef, err := d.l1.L1BlockRefByNumber(ctx, d.origin.Number+1)
+	// temporary error, will do a backoff
+	if err != nil {
+		return err
+	}
+	return d.AdvanceL1Origin(ctx, blkRef.ID())
+}
+
+// Reset the challenge event derivation origin in case of L1 reorg
+func (d *DA) Reset(ctx context.Context, base eth.L1BlockRef, baseCfg eth.SystemConfig) error {
+	// resetting due to expired challenge, do not clear state
```

Comment:
This does not account for other sources of resets. I guess the behavior is when there is a reset triggered by an expired challenge, we want to know that the challenge is still expired when we walk across the data so we need persist that across resets initiated to reorg out those L2 blocks.